agents/tools/upstream_tools.py

```diff
@@ -0,0 +1,702 @@
+from urllib.request import Request, urlopen
+from urllib.error import URLError, HTTPError
+
+from pydantic import BaseModel, Field
+
+from beeai_framework.context import RunContext
+from beeai_framework.emitter import Emitter
+from beeai_framework.tools import JSONToolOutput, StringToolOutput, Tool, ToolError, ToolRunOptions
+
+from common.validators import AbsolutePath
+from utils import run_subprocess
+
+
+class ExtractUpstreamRepositoryInput(BaseModel):
+    upstream_fix_url: str = Field(description="URL to the upstream fix/commit")
+
+
+class UpstreamRepository(BaseModel):
+    """Represents an upstream git repository and commit information."""
+    repo_url: str = Field(description="Git clone URL of the upstream repository")
+    commit_hash: str = Field(description="Commit hash to cherry-pick")
+    original_url: str = Field(description="Original upstream fix URL")
+    pr_number: str | None = Field(default=None, description="Pull request or merge request number if this is a PR/MR URL, None otherwise")
+    is_pr: bool = Field(default=False, description="True if this is a pull request or merge request URL")
+
+
+class ExtractUpstreamRepositoryOutput(JSONToolOutput[UpstreamRepository]):
+    pass
+
+
+class ExtractUpstreamRepositoryTool(Tool[ExtractUpstreamRepositoryInput, ToolRunOptions, ExtractUpstreamRepositoryOutput]):
+    name = "extract_upstream_repository"
+    description = """
+    Extract upstream repository URL and commit hash from a commit or pull request URL.
+
+    Supports common formats:
+    - GitHub/GitLab commit: https://domain.com/owner/repo/commit/hash or /-/commit/hash
+    - GitHub/GitLab PR: https://domain.com/owner/repo/pull/123 or /merge_requests/123
+    - Query param formats: ?id=hash or ?h=hash (for cgit/gitweb)
+
+    For pull requests, fetches the head commit SHA from the PR.
+    Returns the git clone URL and commit hash needed for cherry-picking.
+    """
+    input_schema = ExtractUpstreamRepositoryInput
+
+    def _create_emitter(self) -> Emitter:
+        return Emitter.root().child(
+            namespace=["tool", "upstream", self.name],
+            creator=self,
+        )
+
+    async def _run(
+        self, tool_input: ExtractUpstreamRepositoryInput, options: ToolRunOptions | None, context: RunContext
+    ) -> ExtractUpstreamRepositoryOutput:
+        try:
+            parsed = urlparse(tool_input.upstream_fix_url)
+
+            # Check if this is a pull request URL
+            pr_match = re.search(r'/pull/(\d+)(?:\.patch)?', parsed.path)
+            mr_match = re.search(r'/merge_requests/(\d+)(?:\.patch)?', parsed.path)
+
+            if pr_match or mr_match:
+                # Handle GitHub Pull Request or GitLab Merge Request
+                pr_number = pr_match.group(1) if pr_match else mr_match.group(1)
+
+                # Extract owner/repo from path
+                if pr_match:
+                    repo_path_match = re.match(r'/([\w\-\.]+)/([\w\-\.]+)/pull/', parsed.path)
```

Comment:
Why repeated matching? Wouldn't a single match be enough?

Reply:
I mean, it's not technically wrong, I just think it could be done more effectively (and potentially improving readability).